File: combinationsumii.js
//Objective is to do the same thing as 'Combination Sum' except to only use each
//element only once.

let candidates = [10,1,2,7,6,1,5]
let target = 8


//O() solution

let temp = []
let sum = 0
let index = 0
let result = []

//Sort the elements so we avoid repeats
candidates.sort((a,b) => {
    return a - b
})

backtrack (temp, sum, index)

function backtrack(temp, sum, index) {
    if (sum > target) {
        return
    }

    //Make a copy of the array into the result
    if (sum == target) {
        result.push([...temp])
        return
    }

    for (let i = index; i < candidates.length; i++) {
        if (i !== index && candidates[i] == candidates[i - 1]) {
            continue
        }

        //push the value in then pop it out to make sure it's only used once
        temp.push(candidates[i])
        backtrack(temp, sum + candidates[i], i + 1)
        temp.pop()
    }
}

console.log(result)

File: combinationsumii.java
//Java Solution

class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        int sum = 0;
        int index = 0;
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        LinkedList<Integer> temp = new LinkedList<Integer>();
        
        backtrack(result, temp, sum, index, target, candidates);
        return result;
    }
    
    public void backtrack(List<List<Integer>> result, LinkedList<Integer> temp, int sum, int index, int target, int[] candidates) {
        if (sum > target) {
            return;
        }
        
        if (sum == target) {
            result.add(new ArrayList<Integer>(temp));
            return;
        }
        
        for (int i = index; i < candidates.length; i++) {
            if (i != index && candidates[i] == candidates[i - 1]) {
                continue;
            }
            temp.add(candidates[i]);
            backtrack(result, temp, sum + candidates[i], i + 1, target, candidates);
            temp.removeLast();
        }
    }
}